### DISCUSSION TITLE
Dosing Pumps / Water Pumps for: Aquariums, Hydro/Aquaponic Use Cases

### DISCUSSION THREAD

**Discussion by jmealo** [home-assistant/architecture#1035] [Entity Models]
For aquariums, hydro/aquaponics it'd be useful to have one or more entities representing water pumps and dosing pumps.

I realize this teeters outside of what's considered a smart home; however, folks in these hobbies who use Home Assistant, would consider these part of their smart home.

Using ESPHome, and making some modifications to Home Assistant to better support these devices, you can easily rival the functionality of commercially available solutions.

Most of these are cloud based, and folks who want local control of their smart home, likely want local control of these as well.

The most important entity type to add I believe would be dosing pump, where you can dose a specific amount of a fluid and keep track of the time and amount of the last dose. Referencing the previous value allows for gradual adjustments based on a probe value.

It's often desirable to adjust dosing based on a sensor value (let's say PH).

The dream would be to be able to use the HA UI to set up something like this:

- If PH is stable, dose set amount
- If PH is too high, dose X% of last dose amount
- If PH is too low, dose X% of last dose amount

By tracking the last dose amount, you can make gradual adjustments and avoid any large swings.

I'm not sure if these are exhaustive, I would do more research into what the commercial offerings need in terms of config/parameters:

**Dosing pump**
- Flow rate
- Calibration values
- Last dose

**Pump**
- On / Off
- Adjustable flow (DC pumps)

**Is this SmartHome related / deserving of an HA fork?**
I don't think forking Home Assistant and targeting towards these use cases make sense as the existing temperature, energy monitoring, and light/switch/power control aspects of HA cover the other 80% of what we need to make this work.

Any thoughts? I'd like to implement this in core if possible. I just wanted to get a gut check first :)

**Related discussions:**
- https://github.com/home-assistant/architecture/discussions/943

**Comment by ghost** [comment]
I agree. Some more support for esphome config and calibration inside esphome. 

Doing automation for dosing inside esphome and not triggered by HA just in case there is break in communications. RTC in ESP would be advised. 

Using HA for calibration of pumps inside esphome. Also using HA to alter the dosing schedule.

Something I would like is (I’m terrible is deigning) cards in HA where there is places in pump areas to write what nutrient is being using and then the dosing amounts etc. 

Maybe even a calculator where you input the system volume (400 litres for example) and then the list of nutrients and the amount you need to dose per litre and it’s all worked out. That dosing info then gets sent to esphome.
